File: algorithm/flash-sort.py
"""
Flash Sort Implementation in Python

Flash sort is a distribution-based algorithm that's very fast on uniform data.
It uses a classification step to distribute elements into classes, then sorts each class.

Time Complexity: O(n) average on uniform data, O(n log n) worst case
Space Complexity: O(n)
"""


def flash_sort(arr):
    if len(arr) == 0:
        return arr
    
    n = len(arr)
    sorted_arr = arr.copy()
    
    # Find min and max
    min_val = sorted_arr[0]
    max_val = sorted_arr[0]
    for i in range(1, n):
        if sorted_arr[i] < min_val:
            min_val = sorted_arr[i]
        if sorted_arr[i] > max_val:
            max_val = sorted_arr[i]
    
    if min_val == max_val:
        return sorted_arr
    
    # Classification
    m = int(0.43 * n)
    l = [0] * m
    c1 = (m - 1) / (max_val - min_val)
    
    for i in range(n):
        k = int(c1 * (sorted_arr[i] - min_val))
        l[k] += 1
    
    for k in range(1, m):
        l[k] += l[k - 1]
    
    # Permutation
    hold = sorted_arr[0]
    move = 0
    j = 0
    k = m - 1
    
    while move < n - 1:
        while j > l[k] - 1:
            j += 1
            k = int(c1 * (sorted_arr[j] - min_val))
        
        flash = sorted_arr[j]
        while j != l[k]:
            k = int(c1 * (flash - min_val))
            hold = sorted_arr[l[k] - 1]
            sorted_arr[l[k] - 1] = flash
            flash = hold
            l[k] -= 1
            move += 1
    
    # Insertion sort for final touches
    for i in range(1, n):
        key = sorted_arr[i]
        j = i - 1
        while j >= 0 and sorted_arr[j] > key:
            sorted_arr[j + 1] = sorted_arr[j]
            j -= 1
        sorted_arr[j + 1] = key
    
    return sorted_arr


if __name__ == "__main__":
    arr = [64, 34, 25, 12, 22, 11, 90]
    print("Flash Sort:", flash_sort(arr))

File: algorithm/flash-sort.js
/**
 * Flash Sort Implementation in JavaScript
 * 
 * Flash sort is a distribution-based algorithm that's very fast on uniform data.
 * It uses a classification step to distribute elements into classes, then sorts each class.
 * 
 * Time Complexity: O(n) average on uniform data, O(n log n) worst case
 * Space Complexity: O(n)
 */

function flashSort(arr) {
    if (arr.length === 0) return arr;
    
    const n = arr.length;
    const sorted = [...arr];
    
    // Find min and max
    let min = sorted[0];
    let max = sorted[0];
    for (let i = 1; i < n; i++) {
        if (sorted[i] < min) min = sorted[i];
        if (sorted[i] > max) max = sorted[i];
    }
    
    if (min === max) return sorted;
    
    // Classification
    const m = Math.floor(0.43 * n);
    const l = new Array(m).fill(0);
    const c1 = (m - 1) / (max - min);
    
    for (let i = 0; i < n; i++) {
        const k = Math.floor(c1 * (sorted[i] - min));
        l[k]++;
    }
    
    for (let k = 1; k < m; k++) {
        l[k] += l[k - 1];
    }
    
    // Permutation
    let hold = sorted[0];
    let move = 0;
    let j = 0;
    let k = m - 1;
    
    while (move < n - 1) {
        while (j > l[k] - 1) {
            j++;
            k = Math.floor(c1 * (sorted[j] - min));
        }
        
        const flash = sorted[j];
        while (j !== l[k]) {
            k = Math.floor(c1 * (flash - min));
            hold = sorted[l[k] - 1];
            sorted[l[k] - 1] = flash;
            flash = hold;
            l[k]--;
            move++;
        }
    }
    
    // Insertion sort for final touches
    for (let i = 1; i < n; i++) {
        const key = sorted[i];
        let j = i - 1;
        while (j >= 0 && sorted[j] > key) {
            sorted[j + 1] = sorted[j];
            j--;
        }
        sorted[j + 1] = key;
    }
    
    return sorted;
}

// Example usage
if (require.main === module) {
    const arr = [64, 34, 25, 12, 22, 11, 90];
    console.log('Flash Sort:', flashSort(arr));
}

module.exports = { flashSort };

